### PR TITLE
feat: support sending emails with OnPrem proxy

### PR DESCRIPTION
**Pull request by jjtang1985**
For OnPrem proxy:
1. a socket connection is established, where custom request/response handlers are needed
1. use that socket connection with nodemailer instead of connecting the mail server directly (internet proxy)2. 
1. close the socket + transport 

e2e test:
Application deployed on CF can send mails with OnPrem proxy.

Closes SAP/cloud-sdk-backlog#ISSUENUMBER.

### CODE REVIEW

File: packages/mail-client/src/mail-client.spec.ts [hunk 1]
```diff
@@ -44,6 +56,47 @@ describe('mail client', () => {
...
+      name: 'my-destination',
+      type: 'MAIL',
+      authentication: 'BasicAuthentication',
+      proxyType: 'OnPremise',
+      proxyConfiguration: {
+        host: 'smtp.gmail.com',
+        port: 587
+      }
+    };
+    const mailOptions1: MailOptions = {
```
Comment (tomfrenken) [inline packages/mail-client/src/mail-client.spec.ts:89]: [q] Why is it called `mailOptions1`?
Is there also some other `mailOptions2`?
[req] Either way, if there is `mailOptions2` it should probably have a more specific name.

**Review by tomfrenken — CHANGES_REQUESTED** [review]
Some findings/questions, after that probably a short follow-up call.

File: packages/connectivity/src/scp-cf/connectivity-service-types.ts [hunk 2]
```diff
@@ -32,6 +32,11 @@ export interface ProxyConfiguration {
    * HTTP headers to be added to a request if tunneled through the proxy.
    */
   headers?: ProxyConfigurationHeaders;
+
+  /**
+   * A JWT for proxy authorization.
```
Comment (FrankEssenberger) [inline packages/connectivity/src/scp-cf/connectivity-service-types.ts:37]: [nth] I would assume any authorization header could be accepted by the server. So perhaps adjust the description to something more general: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Proxy-Authorization

Comment (FrankEssenberger) [reply packages/connectivity/src/scp-cf/connectivity-service-types.ts:37]: [q] Also there is a `Proxy-Authorization` property in the `headers` which is added I [here](https://github.com/SAP/cloud-sdk-js/blob/84976bfdc835dd4b4aeadefbc1d3577a33cd43cc/packages/connectivity/src/scp-cf/authorization-header.ts#L144) I thought to the request. Could you double check if you really need this new property.

Comment (FrankEssenberger) [reply packages/connectivity/src/scp-cf/connectivity-service-types.ts:37]: I assume you do not use the destination for the mail, but still we could use the existing property.

File: packages/connectivity/src/scp-cf/connectivity-service.spec.ts [hunk 3]
```diff
@@ -140,4 +145,28 @@ describe('connectivity-service', () => {
       done();
     });
   });
+
+  it('[MAIL] adds a proxy configuration containing at least the host, the port, and the "proxy-authorization"', async () => {
```
Comment (FrankEssenberger) [inline packages/connectivity/src/scp-cf/connectivity-service.spec.ts:149]: [pp] I would use a `descirbe` to separate the `mail` related thest. I do not like the `[MAIL]` tag so much

File: packages/connectivity/src/scp-cf/connectivity-service.ts [hunk 4]
```diff
@@ -65,6 +72,24 @@ export function proxyHostAndPort(): HostAndPort {
...
 }
 
+/**
+ * @internal
+ * @returns Socks Proxy Configuration
+ */
+export async function socksProxyHostAndPort(): Promise<ProxyConfiguration> {
+  const service = readConnectivityServiceBinding();
+  const connectivityServiceCredentials = service.credentials;
+  const connectivityServiceToken = await serviceToken('connectivity', {
```
Comment (FrankEssenberger) [inline packages/connectivity/src/scp-cf/connectivity-service.ts:82]: [req] I think you can pass the service directly to the serviceToken method.

File: packages/connectivity/src/scp-cf/destination/proxy-util.ts [hunk 5]
```diff
@@ -259,6 +259,10 @@ export function proxyAgent(
       return {
         httpsAgent: new HttpsProxyAgent(agentConfig)
       };
+    default:
```
Comment (FrankEssenberger) [inline packages/connectivity/src/scp-cf/destination/proxy-util.ts:262]: [nth] I guess here the name `http-proxy-util.ts`is not more appropiate.

File: packages/mail-client/src/mail-client.spec.ts [hunk 6]
```diff
@@ -44,6 +56,47 @@ describe('mail client', () => {
     expect(spySendMail).toBeCalledTimes(2);
     expect(spySendMail).toBeCalledWith(mailOptions1);
     expect(spySendMail).toBeCalledWith(mailOptions2);
-    expect(spyClose).toBeCalledTimes(1);
+    expect(spyCloseTransport).toBeCalledTimes(1);
+  });
+
+  it('[OnPrem] should create transport/socket, send mails and close the transport/socket', async () => {
```
Comment (FrankEssenberger) [inline packages/mail-client/src/mail-client.spec.ts:62]: puh, Junjie is the master spy -#JamesBond

File: packages/mail-client/src/mail-client.ts [hunk 7]
```diff
@@ -122,7 +177,11 @@ async function sendMailWithNodemailer<T extends MailOptions>(
   mailDestination: MailDestination,
   ...mailOptions: T[]
 ): Promise<MailResponse[]> {
-  const transport = createTransport(mailDestination);
+  let socket: SocksSocket | undefined;
+  if (mailDestination.proxyType === 'OnPremise') {
+    socket = await createSocket(mailDestination);
+  }
+  const transport = await createTransport(mailDestination, socket);
```
Comment (FrankEssenberger) [inline packages/mail-client/src/mail-client.ts:184]: [pp] I know the `for` loop was already there from the last PR. I prefer the pattern of a await Promise.all(opts.map.(opt,idx=>{})) I assume that is also stronger parallelized, but if it is intentional to not overload the server keep it.

File: packages/mail-client/src/mail-client.ts [hunk 8]
```diff
@@ -79,14 +87,50 @@ function getCredentials(
...
+    destination: {
+      host: mailDestination.host!,
+      port: mailDestination.port!
+    }
+  };
+  const socketConnection = await SocksClient.createConnection(
+    connectionOptions
+  );
+  const socksSocket = socketConnection.socket as SocksSocket;
+  socksSocket._readableState.readableListening = true;
```
[the remark this answers is not in the document]
Comment (jjtang1985) [reply packages/mail-client/src/mail-client.ts]: I tested again, and without this change it worked, so I'll remove this line.

File: packages/connectivity/src/scp-cf/connectivity-service-types.ts (hunk 2 above)
Comment (jjtang1985) [reply packages/connectivity/src/scp-cf/connectivity-service-types.ts:37]: Yes, you are right when we are talking about a "header", which saved in line 34.
However, when using non-http (e.g., socket) protocols, the header is not applicable any more.
JWT is a common auth mechanism nowadays, so I added this key.

File: packages/mail-client/src/mail-client.ts (hunk 7 above)
Comment (jjtang1985) [reply packages/mail-client/src/mail-client.ts:182]: I discussed with Marika about the sequential v.s. parallel and we decided to have a small ticket (https://github.com/SAP/cloud-sdk-backlog/issues/820) to make the implementation/test easier.
Once we have the e2e tests in the pipeline, we can send emails in parallel.

**Review by FrankEssenberger — APPROVED** [review]

**Pull request merged**